x/leverage/types/position.go

```diff
@@ -0,0 +1,230 @@
+			sp.LiquidationThreshold, // pairs may not reduce liquidation threshold below what the tokens would produce
+			sdk.MinDec(
+				position.tokenLiquidationThreshold(sp.Collateral),
+				sdk.MaxDec(position.tokenLiquidationThreshold(sp.Borrow), minimumBorrowFactor),
+			),
+		)
+		wp := WeightedSpecialPair{
+			Collateral:           sdk.NewDecCoinFromDec(sp.Collateral, sdk.ZeroDec()),
+			Borrow:               sdk.NewDecCoinFromDec(sp.Borrow, sdk.ZeroDec()),
+			SpecialWeight:        cw,
+			LiquidationThreshold: lt,
+		}
+		// sorting is performed by Add function
+		position.specialPairs = position.specialPairs.Add(wp)
+	}
+
+	for _, cv := range collateralValue {
+		// track total collateral value
+		position.collateralValue = position.collateralValue.Add(cv.Amount)
+	}
+	for _, bv := range borrowedValue {
+		// track total borrowed value
+		position.borrowedValue = position.borrowedValue.Add(bv.Amount)
+	}
+
+	// match assets into special asset pairs, removing matched value from collateralValue and borrowedValue
+	for _, p := range position.specialPairs {
+		b := borrowedValue.AmountOf(p.Borrow.Denom)
+		c := collateralValue.AmountOf(p.Collateral.Denom)
+		if b.IsPositive() && c.IsPositive() {
+			// some unmatched assets match the special pair
+			pairBorrowLimit := c.Mul(p.SpecialWeight)
+			if pairBorrowLimit.GTE(b) {
+				// all of the borrow is covered by collateral in this pair
+				borrowedValue = borrowedValue.Sub(sdk.NewDecCoins(sdk.NewDecCoinFromDec(p.Borrow.Denom, b)))
+				collateralValue = collateralValue.Sub(sdk.NewDecCoins(sdk.NewDecCoinFromDec(
+					// some collateral, equal to borrow value / collateral weight, is used
+					p.Collateral.Denom, b.Quo(p.SpecialWeight),
+				)))
+			} else {
+				// only some of the borrow, equal to collateral value * collateal weight is covered
+				borrowedValue = borrowedValue.Sub(sdk.NewDecCoins(sdk.NewDecCoinFromDec(
+					p.Borrow.Denom, c.Mul(p.SpecialWeight),
+				)))
+				collateralValue = collateralValue.Sub(sdk.NewDecCoins(sdk.NewDecCoinFromDec(p.Collateral.Denom, c)))
+			}
+		}
+	}
+
+	for _, cv := range collateralValue {
+		// sort collateral by collateral weight using Add function
+		position.collateral = position.collateral.Add(
+			WeightedDecCoin{
+				Asset:       sdk.NewDecCoinFromDec(cv.Denom, cv.Amount),
+				Weight:      position.tokenCollateralWeight(cv.Denom),
+				Liquidation: position.tokenLiquidationThreshold(cv.Denom),
+			},
+		)
+	}
+
+	for _, bv := range borrowedValue {
+		// sort borrows by collateral weight using Add function
+		position.borrowed = position.borrowed.Add(
+			WeightedDecCoin{
+				Asset:       sdk.NewDecCoinFromDec(bv.Denom, bv.Amount),
+				Weight:      position.tokenCollateralWeight(bv.Denom),
+				Liquidation: position.tokenLiquidationThreshold(bv.Denom),
+			},
+		)
+	}
+
+	return position
+}
+
+// BorrowedValue returns an account's total USD value borrowed
+func (ap *AccountPosition) BorrowedValue() sdk.Dec {
+	return ap.borrowedValue
+}
```

Comment:
Keeping all the fields on this struct private to prevent accidental misuse: this value should be immutable once the struct is initialized, so we don't expose it to be set from the keeper package (or others)